core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -227,7 +231,7 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
     } finally {
       running.set(false)
 
-      if (RuntimeFlags.currentFiber(_runtimeFlags)) Fiber._currentFiber.set(null)
+      if (RuntimeFlags.currentFiber(_runtimeFlags)) Fiber._currentFiber.set(previousFiber)
```

Comment:
I don't think this is quite right: because what if the runtime flags change between the start and the end block?

Maybe something like:

```suggestion
      if ((previousFiber ne null) || RuntimeFlags.currentFiber(_runtimeFlags)) Fiber._currentFiber.set(previousFiber)
```

Reply:
Agreed. If we set the current fiber we need to restore it even if the flag is disabled.